File: 6236.py
import sys

n, m = map(int, sys.stdin.readline().split())
arr = [0]*n

for i in range(n):
    arr[i] = int(sys.stdin.readline())
    
low = min(arr)
high = sum(arr)

while low <= high:
    mid = (low+high)//2
    cnt = 1
    cash = mid

    for i in arr:
        if cash < i:
            cnt += 1
            cash = mid
        cash -= i

    if cnt > m or mid < max(arr):
        low = mid+1        
    else:
        high = mid-1
        k = mid

print(k)

File: 6236.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main(){

    ios::sync_with_stdio(false);
    cin.tie(0);
    
    int n, m;
    cin >> n >> m;
    
    vector<int> arr(n);
    int low = 0, high = 0;
    
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
        low = max(low, arr[i]);  
        high += arr[i];
    }
    
    int result = high;
    
    while (low <= high) {
        int mid = (low + high) / 2;
        int cnt = 1, cash = mid;
        
        for (int i = 0; i < n; i++) {
            if (cash < arr[i]) {
                cnt++;
                cash = mid;
            }
            cash -= arr[i];
        }
        
        if (cnt > m) {
            low = mid + 1;
        } else {
            result = mid;
            high = mid - 1;
        }
    }
    
    cout << result << "\n";
    return 0;
}